### PR TITLE
Fix race condition in sonobuoy logs

### LINKED ISSUE #348: logs is truncating output occasionally
sometimes it prints all the logs, sometimes it doesn't and the last line looks like this:
```
$ sonobuoy logs --kubeconfig aks-test-kubeconfig 
...
time="2018-03-08T20:13:15Z" level=error msg="error querying CronJobs: the server could not find the reqaks salazar:sonobuoy cha$ 
```

### PR DESCRIPTION
**Pull request by chuckha**
Closes #348

Signed-off-by: Chuck Ha <chuck@heptio.com>

### CODE REVIEW

File: pkg/client/logs_test.go [hunk 1]
```diff
@@ -67,20 +68,20 @@ func TestLateErrors(t *testing.T) {
 
 func TestLogEarlyErrors(t *testing.T) {
 	input := "sonobuoy will help you on your way to greatness"
-	bytestream := make(chan []byte, 1)
-	bytestream <- []byte(input)
-
-	errc := make(chan error, 1)
-	errc <- errors.New("A seriously bad error")
+	bytestream := make(chan []byte)
+	go func() {
```
Comment (davecheney) [inline pkg/client/logs_test.go]: this is the same as writing

bytestream := make(chan []byte, 1)
bytestream <- []byte(input)

File: pkg/client/logs.go [hunk 2]
```diff
@@ -83,30 +86,25 @@ func (r *Reader) Read(p []byte) (int, error) {
...
-		}
-
-		// The incoming data is bigger than size of the remaining size of the buffer. Save overflow data for next read.
-		remainingSize := len(p) - n
-		if len(data) > remainingSize {
-			copy(p[n:], data[:remainingSize])
-			r.overflowBuffer = data[remainingSize:]
-			return len(p), nil
-		}
+	data, ok := <-r.bytestream
```
Comment (davecheney) [inline pkg/client/logs.go]: line 84, r.overflowBuffer = nil

File: pkg/client/logs_test.go (hunk 1 above)
Comment (chuckha) [reply pkg/client/logs_test.go]: not sure what you'd like me to do with this comment, can you give me a little more context? do you find this easier to read?

File: pkg/client/logs.go (hunk 2 above)
Comment (chuckha) [reply pkg/client/logs.go]: ah sweet, i knew there was an easier way

File: pkg/client/logs.go [hunk 3]
```diff
@@ -78,35 +81,29 @@ func (r *Reader) Read(p []byte) (int, error) {
...
+	remainingSize := len(p) - n
+	if len(data) > remainingSize {
+		copy(p[n:], data[:remainingSize])
+		r.overflowBuffer = data[remainingSize:]
+		return len(p), nil
 	}
+
+	// We have enough headroom in the buffer, copy all of it.
+	copy(p[n:], data)
+	n += len(data)
```
Comment (liztio) [inline pkg/client/logs.go]: should just be `n = len(data)`

Comment (liztio) [reply pkg/client/logs.go]: delete n, just return `len(data)`

**Pull request merged**